include/stl2/detail/to.hpp

```diff
@@ -0,0 +1,168 @@
+//
+#ifndef STL2_DETAIL_TO_HPP
+#define STL2_DETAIL_TO_HPP
+
+#include <stl2/detail/closure.hpp>
+#include <stl2/detail/fwd.hpp>
+#include <stl2/detail/meta.hpp>
+#include <stl2/detail/iterator/common_iterator.hpp>
+#include <stl2/detail/range/access.hpp>
+#include <stl2/detail/range/concepts.hpp>
+
+STL2_OPEN_NAMESPACE {
+	namespace __to {
+		template<class _Cont>
+		META_CONCEPT __containerish =
+			ForwardRange<_Cont> /* && !View<_Cont> */ && MoveConstructible<_Cont>;
+
+		template<class _Cont, class _IRange, class... _Args>
+		META_CONCEPT __class_convertible = InputRange<_IRange> &&
+			__containerish<std::remove_cv_t<_Cont>> &&
+			Constructible<ext::range_value_t<_Cont>, ext::range_reference_t<_IRange>> &&
+			Constructible<_Cont, __range_common_iterator<_IRange>,
+				__range_common_iterator<_IRange>, _Args...>;
+
+		template<class _IRange, class _Cont, class... _Args>
+		META_CONCEPT __reserve_assignable = SizedRange<_IRange> && Constructible<_Cont, _Args...> &&
+			requires(_Cont& __mc, const _Cont& __cc) {
+				{ __cc.capacity() } -> Integral;
+				__mc.reserve(__cc.capacity());
+				requires Assignable<ext::range_reference_t<_Cont>, ext::range_reference_t<_IRange>>;
+				__mc.assign(__range_common_iterator<_IRange>(), __range_common_iterator<_IRange>());
+			};
+
+		template<class _Cont>
+		struct __class_to {
+			template<class _IRange, class... _Args>
+			requires __class_convertible<_Cont, _IRange, _Args...>
+			constexpr _Cont operator()(_IRange&& __r, _Args&&... __args) const {
+				using __ci = __range_common_iterator<_IRange>;
+				if constexpr (__reserve_assignable<_IRange, _Cont, _Args...>) {
+					_Cont __c(static_cast<_Args&&>(__args)...);
+					auto const __size = size(__r);
+					using __capacity = decltype(__c.capacity());
+					// if (__can_represent<__capacity>(__size)) ?
+					__c.reserve(static_cast<__capacity>(__size));
+					__c.assign(__ci(begin(__r)), __ci(end(__r)));
+					return __c;
+				} else {
+					return _Cont(__ci(begin(__r)), __ci(end(__r)), static_cast<_Args&&>(__args)...);
+				}
+			}
+		};
+
+		template<class _Cont, class... _Args>
+		requires (CopyConstructible<_Args> && ...)
+		struct STL2_EMPTY_BASES __class_closure : detail::closure<__class_to<_Cont>, _Args...> {
+			constexpr __class_closure(_Args&&... __args)
+			: detail::closure<__class_to<_Cont>, _Args...>{{}, static_cast<_Args&&>(__args)...} {}
+		};
+
+		template<class _Cont>
+		constexpr __class_closure<_Cont> to() noexcept {
+			return {};
+		}
+
+		template<class _IRange, class _Cont>
+		requires __class_convertible<_Cont, _IRange>
+		constexpr auto operator|(_IRange&& __r, __class_closure<_Cont>(&)() noexcept) {
+			return __class_to<_Cont>{}(static_cast<_IRange&&>(__r));
+		}
+
+		template<class _Cont, class _First, class... _Rest>
+		requires __class_convertible<_Cont, _First, _Rest...> ||
+			(CopyConstructible<_First> && (CopyConstructible<_Rest> && ...))
+		constexpr auto to(_First&& __first, _Rest&&... __rest) {

```

Comment:
I'm curious how you expect the extra arguments to be used.

Reply:
The examples in the paper / the unit tests use them to pass allocators; I'm not aware of any other stated use cases.